Fix revisiting order confirmation and guest to customer conversion in FO

Hlavtox opened the pull request:
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | See below
| Type?             | bug fix
| Category?         | FO
| BC breaks?        | no
| Deprecations?     | yes
| Fixed ticket?     | Fixes #14414 and #20388 and #23230
| How to test?      | See below
| Possible impacts? | -

## Description
- I rewrote and improved order confirmation controller inner workings. This allowed to fix several things, made things cleaner and more bulletproof.
- ✔️ When you refreshed order confirmation page, you got redirected to login screen or order history. This is now working 100% properly.
- ✔️ When you wanted to convert your account to customer, you had to enter many things and it didn't work at all.
- ✔️ When you enabled catalog mode after making an order, you could no longer access the confirmation screen. This is now working.

## Technical details
- Email on confirmation screen is now taken from the order itself, not from the context. So you can checkout an order in your browser history you made from a different account, and still see the correct information.
- There is a check removed on id_customer, that prevented the revisiting the page, because guests are "logged out" from cookie on first confirm visit. This check is not needed. Mechanism that prevents customer to see other customer's order is the secure_key.
- The PR is prepared for another PR, that will fix ordering as a guest with customer existing - https://github.com/PrestaShop/PrestaShop/pull/25353.
- I removed check on Catalog mode. If you enable catalog mode after customers have been ordering, why they can't view their order confirmation?

## Guest to customer conversion
- The box is shown only to guests and only if a customer account with the same email does not exist.
- They now only need to enter the password (like on guest tracking page), not the name and email again.
- There is a nice error/notification for all cases - empty password, short password, on page refresh, if they already converted him in BO.
- I unified both forms on confirmation page and guest tracking page to a single TPL.
- I added a check for empty password and already converted customer also to guest tracking controller.
- I renamed a variable I added in my previous PR, so it's unified.
![Výstřižek](https://user-images.githubusercontent.com/6097524/126567914-2d6acedd-62b2-4f74-8928-62ac4a2ba6ca.JPG)


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/25395)
<!-- Reviewable:end -->

Hlavtox committed:
Fix revisiting order confirmation and guest to customer conversion

kpodemski commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -45,31 +48,92 @@ public function init()
-
-        $redirectLink = 'index.php?controller=history';
-
+        // We get the required data from URL
```
> I am pretty rich on coments, if it's not a good practice, I will reduce it, but this way it can be read like a book.

Actually, it should be a code that you read like a book! :) For example in this place, it's obvious that we're getting data from an URL, same for checking password length, etc.

matks replied:
Whether comments are good or not has been a discussion for dozen of years among developers.

Famous quote by Robert C Martin (author of Clean Code):
> A comment is a failure to express yourself in code. If you fail, then write a comment; but try not to fail.

But plenty of people, including others renowned authors have disagreed with him.

I don't have the Ultimate Truth but I think @kpodemski opinion is a good one: comments should bring some informations to the reader that he could not find easily from the code. If you can grasp an information either from the code (getting data from an URL) or from a comment, the comment is not useful. Look at @jolelievre comments in PR https://github.com/PrestaShop/PrestaShop/pull/24991/ they give such information.

However, when you have to choose whether to add a comment or not to add, I think it's always good to add it :) it's better to have a code that is uselessly too obvious rather than an obscure code 😄 

That being said... I'm quite sure some of the other maintainers will disagree with me :trollface:

eternoendless replied:
I agree that it's preferable to risk having redundant comments rather than lose valuable contextual information. Don't make me reverse engineer code!

kpodemski commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -79,31 +143,26 @@ public function init()
      */
     public function initContent()
     {
-        if (Configuration::isCatalogMode()) {
```
I think this check was on point?

Hlavtox replied:
I removed this, because imagine a situation - You have catalog mode disabled and ordering enabled. Somebody makes an order, you enable catalog mode 5 minutes after that. Why would he not be able to re-check his order confirm? :-)

kpodemski commented on themes/classic/templates/checkout/order-confirmation.tpl:
```diff
@@ -93,11 +93,31 @@
+            <li> -{l s='Fast and easy checkout' d='Shop.Theme.Customeraccount'}</li>
+            <li> -{l s='Easier merchandise return' d='Shop.Theme.Customeraccount'}</li>
+          </ul>
+          <form method="post">
```
I'm not in favor of hardcoding theme forms, it's really good for developer to handle form markup in one place instead of looking for forms all over the place. Consider building this form in the same way like we build login form etc.

Hlavtox replied:
@kpodemski It's nearly the same hardcoded form as on guest tracking. I think we could merge it into one TPL and reuse it, at least.

Progi1984 commented:
The linked issue #20388 wait for @PrestaShop/product-team answer :)

matks commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -45,31 +48,92 @@ public function init()
-        $order = new Order((int) ($this->id_order));
+        $this->order = new Order((int) ($this->id_order));
 
+        // We run all the required checks and if something is smelly, we redirect him to order history
```
One way to avoid having to write a comment is to use variable names
For example
```
        // We run all the required checks and if something is smelly, we redirect him to order history
```
can transform into
```
$somethingIsSmelly = (!Validate::isLoadedObject($this->order) || $this->secure_key != $this->order->secure_key);
if ($somethingIsSmelly)
```

eternoendless replied:
It's usually better to have dedicated method

```php
if (!$this->verifyOrder($this->order, $this->secure_key)) {
    // error
}
```

matks commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -45,31 +48,92 @@ public function init()
+        $this->id_order = (int) (Tools::getValue('id_order', 0));
         $this->secure_key = Tools::getValue('key', false);
-        $order = new Order((int) ($this->id_order));
+        $this->order = new Order((int) ($this->id_order));
```
I think this change is important. `$order` ⏩  `$this->order` can you explain it ?

Hlavtox replied:
@matks It's because previously, the Order was created twice for nothing - `$order = new Order(Order::getIdByCartId((int) ($this->id_cart)));`. Now I initialize the object only once in init() and save it to the attribute, so we can use it in all other methods.

matks commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -45,31 +48,92 @@ public function init()
+                    'Shop.Forms.Help'
+                );
+
+            // If the password is short
```
These comments are not needed, `(strlen($password) < Validate::PASSWORD_LENGTH)` is exactly "verify password is long enough"

eternoendless commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -23,18 +23,21 @@
-    public $reference;
     public $secure_key;
-    public $order_presenter;
+    public $order;
+    public $customer;
```
Consider making these protected

Hlavtox commented:
@matks @eternoendless @kpodemski 
- Ensured compatibility by preserving `id_cart` and `reference` attributes, if anyone is overriding this. New users can use `order` attribute.
- Cleaned unneeded comments and made the two new attributes protected.
- Removed some old historic code `slowvalidation`. It seems it was only used like in 2010 in PayPal module. No trace of it in last years anywhere.

Matt75 commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -45,31 +48,86 @@ public function init()
+                    'Shop.Notifications.Success'
+                );
+            } else {
+                $this->success[] = $this->trans(
```
```suggestion
                $this->errors[] = $this->trans(
```

Matt75 commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -45,31 +48,86 @@ public function init()
         }
-        $this->reference = $order->reference;
-        if (!Validate::isLoadedObject($order) || $order->id_customer != $this->context->customer->id || $this->secure_key != $order->secure_key) {
+        $redirectLink = 'index.php?controller=history';
```
```suggestion
        $redirectLink = $this->context->link->getPageLink('history', $this->ssl);
```
Avoid double redirection in case of Url Rewrite enabled

Hlavtox commented:
@Matt75 Thank you man, all applied! 🚀

Matt75 commented:
There is a problem with using id_order instead of id_cart because after payment a Cart can be split into one or more orders.

Multishipping feature has been partially removed in 1.6.1 and moreover in 1.7.0 but code is still there in PaymentModule class.

And some conditions can currently split a Cart into multiple orders : https://github.com/PrestaShop/prestashop-specs/issues/111

That's why we use id_cart instead of id_order because before 1.7.0 we display all orders created with this Cart.

1.7.0 introduced a regression with confirmation page because only one order is displayed on confirmation page : https://github.com/PrestaShop/PrestaShop/issues/9594

It's very confusing for a customer to pay for 5 products for example and confirmation page show only 3 products with a different total price... its a real problem

ldeprestashop commented:
We are looking (not date fix, we must prioritize many features)  for multishipping for future version that's why I agree with @Matt75 last comment.

Hlavtox commented:
@Matt75 @ldeprestashop 

Hi guys, I didn't know this was possible. I think many modules are not even ready for this, but we can for sure fix it!

However, I checked 1.6.0.X code and I couldn't find anything that would display two orders on confirmation screen, are you sure it was there?

1) Can you provide me with how I can reproduce this multishipping?
2) Do you want me to do it in this PR or we can merge this (I will revert to using id_cart) and do it in another PR?
3) Do you have some specs for this? I would change order attribute to orders -> do what I want with them -> In the template, if there are two orders, I will display message like “We split your cart into multiple orders (ABCDFG and HIJKLMN), so we can work on them independently.” and two product tables.

Matt75 commented:
@Hlavtox

Test scenarios in https://github.com/PrestaShop/prestashop-specs/issues/111 with older versions than PrestaShop 1.7 and you should see all related orders created from cart paid.

1. Multishipping can only be reproduced with PrestaShop 1.5 and carrier modules compliant was rare but others cases bellow can create multiple orders.

For others questions, I let Product Team answer

Hlavtox commented:
@Matt75 Sorry man, but I don't want to do it in this PR. I will keep the way of getting order from cart id, but let's extend it in another PR. It's too complicated and it would be a mess.

Matt75 commented:
> @Matt75 Sorry man, but I don't want to do it in this PR. I will keep the way of getting order from cart id, but let's extend it in another PR. It's too complicated and it would be a mess.

No worries, my goal was just to explain everyone why id_cart is used and why use id_order instead will cause more problem with current situation

Moreover we need a decision to choose witch behavior is good

marionf commented:
Hello @Hlavtox 

Thank you for this really nice PR ! 

I am ok with all the behaviors you described in the PR ✔️  
However, while testing it I ran into the following error when I tried to sign in after the guest account was transformed into a customer through the FO:
<img width="929" alt="Capture d’écran 2021-08-18 à 11 25 29" src="https://user-images.githubusercontent.com/13449658/129874446-35b81ddd-136c-422f-8ff8-c181df2c2fc8.png">

I also found another issue:
1) I made an order as a guest
2) I transformed the guest into a customer account through the back-office
3) In FO, on the order-confirmation page (I did not refresh it), I tried to enter a password and click on "Create account"
4) It says "Your guest account has been already transformed into a customer account. You can log in as a registered shopper.
"
5) I tried to sign in to my customer account but the password is not working.

I think we should display an error message if the guest was already transformed into a customer account while clicking on "Create account".

Hlavtox commented:
@marionf Hi Marion, the first bug is fixed, I put a variable onto wrong line. Ooops :-)

Regarding the second situation - I changed the message to error and improved the wording. So he knows that the conversion with his password did not went through. It's easy solution not requiring other mods and logic. I think it should be clear now.

![Výstřižek](https://user-images.githubusercontent.com/6097524/130028521-d6c78e93-849d-491d-b5d0-848783abb4c1.JPG)

_______

@Matt75 Reverted the logic to getting the data by id_cart so it's prepared for order splitting. 👍

Julievrz commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -45,31 +48,94 @@ public function init()
+
+            if (empty($password)) {
+                $this->errors[] = $this->trans(
+                    'To convert your account, you must enter a password.',
```
```suggestion
                    'Enter a password to transform your guest account into a customer account.',
```

Julievrz commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -45,31 +48,94 @@ public function init()
+            // B) if we already transformed him in other window or through backoffice
+            } elseif ($this->customer->is_guest == 0) {
+                $this->errors[] = $this->trans(
+                    'We could not transform your guest account, it has already been transformed into a customer account. You can log in as a registered shopper.',
```
```suggestion
                    'A customer account has already been created from this guest account. Please sign in.',
```
If I understand well, this is what you mean here? ☺️

Julievrz commented on themes/classic/templates/checkout/order-confirmation.tpl:
```diff
@@ -93,11 +93,31 @@
         <div class="card-block">
-          <h4 class="h4">{l s='Save time on your next order, sign up now' d='Shop.Theme.Checkout'}</h4>
-          {render file='customer/_partials/customer-form.tpl' ui=$register_form}
+          <h4 class="h4">{l s='Save time on your next order, sign up now and enjoy:' d='Shop.Theme.Checkout'}</h4>
```
```suggestion
          <h4 class="h4">{l s='Save time on your next order, sign up now' d='Shop.Theme.Checkout'}</h4>
```
"and enjoy:" followed by bullet points is not really easy to translate.

Hlavtox commented:
@Julievrz Wording updated ;-)

Hlavtox commented:
@matks @eternoendless Hi guys, I unified the conversions into one file and improved error messages on guest tracking. :-)

Progi1984 commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -23,18 +23,21 @@
-    public $reference;
     public $secure_key;
-    public $order_presenter;
+    protected $order;
+    protected $customer;
+    public $reference; // Deprecated
+    public $id_cart; // Deprecated
+    public $order_presenter; // Deprecated
```
Could you add PHPDoc for these attributes ?

kpodemski commented on controllers/front/GuestTrackingController.php:
```diff
@@ -129,7 +145,7 @@ public function initContent()
         $this->context->smarty->assign([
             'order' => $presented_order,
             'guest_email' => Tools::getValue('email'),
-            'is_customer' => Customer::customerExists(Tools::getValue('email'), false, true),
```
To avoid a big BC Break I'd leave it as a deprecated variable

kpodemski commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -79,31 +145,26 @@ public function init()
-            ->makeCustomerForm()
-            ->setGuestAllowed(false)
-            ->fillWith(Tools::getAllValues());
+        // Initialize presenters
```
```suggestion
```

kpodemski commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -79,31 +145,26 @@ public function init()
 
         parent::initContent();
 
+        // Assign data to output and set template
```
```suggestion
```

kpodemski commented on themes/classic/templates/checkout/order-confirmation.tpl:
```diff
@@ -92,16 +92,15 @@
         <div class="card-block">
-          <h4 class="h4">{l s='Save time on your next order, sign up now' d='Shop.Theme.Checkout'}</h4>
-          {render file='customer/_partials/customer-form.tpl' ui=$register_form}
+          {include file='customer/_partials/conversion-form.tpl'}
```
Naming is a bit awkward for me, as conversion has a completely different meaning most of the time. What do you think about `account-transformation-form.tpl`?

kpodemski commented on themes/classic/templates/customer/_partials/conversion-form.tpl:
```diff
@@ -0,0 +1,46 @@
+      <li> - {l s='Fast and easy checkout' d='Shop.Theme.Customeraccount'}</li>
+      <li> - {l s='Easier merchandise return' d='Shop.Theme.Customeraccount'}</li>
+    </ul>
+    <form method="post">
```
you could add a unique id based on the template name

Hlavtox replied:
@kpodemski How do you want me to do it?

Progi1984 commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -45,31 +48,94 @@ public function init()
     {
         parent::init();
 
+        // If are coming to this page to finish free order, we need to run additional logic
```
```suggestion
        // If we are coming to this page to finish free order, we need to run additional logic
```

Progi1984 commented on controllers/front/OrderConfirmationController.php:
```diff
@@ -45,31 +48,94 @@ public function init()
 
+        /*
+         * Because of order splitting scenarios, we must get the data by id_cart parameter (not id_order),
+         * so we can display all orders made fron this cart.
```
```suggestion
         * so we can display all orders made from this cart.
```

Progi1984 commented on themes/classic/templates/customer/_partials/conversion-form.tpl:
```diff
@@ -0,0 +1,46 @@
+      <li> - {l s='Easier merchandise return' d='Shop.Theme.Customeraccount'}</li>
+    </ul>
+    <form method="post">
+      <div>
```
Can we remove this `div` ?

Progi1984 reviewed (commented):
@Hlavtox Thank you for your contribution. I have brought some comments to your PR.

Hlavtox commented:
@Progi1984 @kpodemski Thanks for checking guys, all suggestions applied. ;-)

kpodemski commented:
@Hlavtox we need wording approval from @Julievrz and then ready for QA 👍🏻

Julievrz commented on controllers/front/GuestTrackingController.php:
```diff
@@ -81,12 +81,28 @@ public function postProcess()
+                $this->errors[] = $this->trans(
+                    'A customer account has already been created from this guest account. Please sign in.',
+                    [],
+                    'Shop.Notifications.Success'
```
```suggestion
                    'Shop.Notifications.Error'
```

Sorry for the delay, I was out of the office last week. I just missed this domain. It's actually an error, not a success. :-)

Hlavtox replied:
@Julievrz No worries. Good catch, completely forgot to check it :). I also fixed it in confirmation controller. 👍

Progi1984 commented:
@Hlavtox You have some conflicting file. Could you rebase your PR, please ?

Hlavtox committed:
Update OrderConfirmationController.php

hibatallahAouadni commented:
Hello @Hlavtox 

Thanks for your PR!
LGTM, QA :heavy_check_mark: 
See the attached screen record below:
https://watch.screencastify.com/v/H56ACdIEzeBCNXM1g3EA

Thanks!

matks commented:
Thank you @Hlavtox

Hlavtox commented:
@matks Thanks Mathieu! :-)
